editor/connections_dialog.cpp

```diff
@@ -517,16 +517,41 @@ String ConnectDialog::get_signature(const MethodInfo &p_method, PackedStringArra
 		}
 
 		const PropertyInfo &pi = p_method.arguments[i];
-		String tname = "var";
-		if (pi.type == Variant::OBJECT && pi.class_name != StringName()) {
-			tname = pi.class_name.operator String();
-		} else if (pi.type != Variant::NIL) {
-			tname = Variant::get_type_name(pi.type);
+		String type_name;
+		switch (pi.type) {
+			case Variant::NIL:
+				type_name = "Variant";
+				break;
+			case Variant::INT:
+				if ((pi.usage & PROPERTY_USAGE_CLASS_IS_ENUM) && pi.class_name != StringName() && !String(pi.class_name).begins_with("res://")) {
+					type_name = pi.class_name;
+				} else {
+					type_name = "int";
+				}
+				break;
+			case Variant::ARRAY:
+				if (pi.hint == PROPERTY_HINT_ARRAY_TYPE && !pi.hint_string.is_empty() && !pi.hint_string.begins_with("res://")) {
+					type_name = "Array[" + pi.hint_string + "]";
```

Comment:
For `PROPERTY_HINT_ARRAY_TYPE`, isn't the hint string formatted as `%s/%s:%s`? See the [`MAKE_RESOURCE_TYPE_HINT`](https://github.com/godotengine/godot/blob/3ed4497113fa10611b90290ce22a751fb9d26e2e/core/object/object.h#L142-L144) macro.

Reply:
> What about `PROPERTY_HINT_TYPE_STRING`?

I think `PROPERTY_HINT_TYPE_STRING` is mainly for the editor. The [core](https://github.com/godotengine/godot/blob/3ed4497113fa10611b90290ce22a751fb9d26e2e/core/variant/typed_array.h#L154-L188), documentation ([one](https://github.com/godotengine/godot/blob/3ed4497113fa10611b90290ce22a751fb9d26e2e/core/doc_data.cpp#L58-L59), [two](https://github.com/godotengine/godot/blob/3ed4497113fa10611b90290ce22a751fb9d26e2e/editor/doc_tools.cpp#L523-L524)) and GDScript ([one](https://github.com/godotengine/godot/blob/3ed4497113fa10611b90290ce22a751fb9d26e2e/modules/gdscript/gdscript_analyzer.cpp#L4749), [two](https://github.com/godotengine/godot/blob/3ed4497113fa10611b90290ce22a751fb9d26e2e/modules/gdscript/gdscript_parser.cpp#L4392)) use `PROPERTY_HINT_ARRAY_TYPE`.

> See the [`MAKE_RESOURCE_TYPE_HINT`](https://github.com/godotengine/godot/blob/3ed4497113fa10611b90290ce22a751fb9d26e2e/core/object/object.h#L142-L144) macro.

This is a bug. See #77618 and https://github.com/godotengine/godot/pull/81293#discussion_r1316095969.

Reply:
Sorry, I'm a bit confused. Are you saying exported properties should use `PROPERTY_HINT_TYPE_STRING` and method/signal parameters should use `PROPERTY_HINT_ARRAY_TYPE`? I don't really understand when to use one over the other, they seem to serve the same purpose at least when type is `Variant:ARRAY`.

I'm trying to understand this because C# method/signal parameters currently use `PROPERTY_HINT_NONE` (so arrays remain untyped even after the changes in this PR[^1]), and I was wondering if we could reuse the code we use when generating the `hint` and `hint_string` for exported members. But exported array properties use `PROPERTY_HINT_TYPE_STRING` right now (I believe [GDScript](https://github.com/godotengine/godot/blob/3ed4497113fa10611b90290ce22a751fb9d26e2e/modules/gdscript/gdscript_parser.cpp#L4128-L4130) does too for exported properties).

[^1]: To clarify, your PR makes generating signal callbacks in C# scripts work as expected as long as the signal is built-in or declared in a GDScript. But signals declared in C# scripts still use untyped arrays, because the parameter `PropertyInfo` always uses `PROPERTY_HINT_NONE`.

Reply:
I agree that we do have some confusion. Note that “typed” arrays appeared in the Inspector earlier than in the core. You can use untyped arrays, and encode in `hint_string` the element type, as well as `hint` and `hint_string` of the elements. The Inspector currently does not check the `hint` of the array at all, only the `hint_string`. This is probably where the confusion arose.

With typed arrays, `PROPERTY_HINT_ARRAY_TYPE` was added to the core, `hint_string` expecting an element type as a string, no way to specify `hint` and `hint_string` of elements. However, support for `PROPERTY_HINT_ARRAY_TYPE` was not implemented in the Inspector (see #77618).

`PROPERTY_HINT_ARRAY_TYPE` is used in the stricter typed array scenario: in the core, documentation, and GDScript when it receives information from the core. As for `PROPERTY_HINT_TYPE_STRING`, perhaps its use with the older (and more functional) `hint_string` format was not intended (since the editor accepts any `hint`). However, GDScript already used `PROPERTY_HINT_TYPE_STRING` to export typed arrays, and the documentation for this constant contained slightly incorrect information. So in #77734 I documented it properly.

By the way, the inability to represent _any_ type using `PropertyInfo` (for example, a non-global GDScript class, inner class, GDScript created on the fly) causes certain inconvenience. We should think about this. Maybe we should add a new type `GodotType` to represent a type. This would make it easier to add nested types (`Array[Array[int]]`), generic types, allow you to switch the type hint format in the editor depending on the language, contextualize the type chain (`MyClass.InnerClass.MoreInnerClass`) and/or namespaces (see godotengine/godot-proposals#6502).